enhancements/security/x509-trust.md

```diff
@@ -0,0 +1,213 @@
+* `composition`, configuring how the bundle in `ca-bundle.crt` is composed with the component's default trust store.
+    Valid values are:
+    * `union`, in which case components should use the union of the `ca-bundle.crt` and their default trust store.
+    * `override`, in which case components should use only `ca-bundle.crt` and ignore their default trust store.
+        This behavior is also the default when a component recieves an unrecognized `composition` value.
+
+### Per-component Trust
+
+Each component will define its own well-known ConfigMap(s) in its own namespace, with the same `ca-bundle.crt` and `composition` keys as [the cluster-scoped ConfigMap](#cluster-scoped-default-trust).
+For example, the registry operator uses [`trusted-ca`][registry-configmap-name] in [the `openshift-image-registry` namespace][registry-configmap-namespace].
+Components that have distinct trust domains may define multiple ConfigMaps for each domain.
+For example, there may be one ConfigMap for proxy/egress TLS, and another ConfigMap for [S/MIME][smime] verification.
+To populate that trust bundle, the cluster administrator can set labels on the ConfigMap(s).
+For example, [the `config.openshift.io/inject-proxy-cabundle` label](../proxy/global-cluster-egress-proxy.md#implementation-detailsnotesconstraints-optional) asks for the current proxy trust bundle.
+
+The cluster-version operator (CVO) [merges ConfigMaps][cluster-version-operator-EnsureConfigMap] by [clobbering any manifest-defined data][cluster-version-operator-mergeMap], [labels, and annotations][cluster-version-operator-EnsureObjectMeta].
+Data keys, labels, and annotations not defined in the manifest are ignored.
+This means that components which decide to provide their per-component ConfigMaps via the CVO should not set labels like `config.openshift.io/inject-default-cabundle` unless they also set [the `release.openshift.io/create-only` annotation to `true`][cluster-version-operator-create-only], to allow cluster administrators to manipulate the per-component ConfigMap without the CVO stomping on their updates.
+For example, the registry operator [sets `create-only`][registry-configmap-create-only].
+
+#### Universal Base Image Containers
+
+FIXME: UBI-specific docs for this [RHEL behavior](rhel-cert-injection) that show that you don't need to install some non-UBI RHEL stuff to make it work.
+
+Components can mount `ca-bundle.crt` from their ConfigMap(s) under `/usr/share/pki/ca-trust-source/anchors/`.
+Components are then responsible for running `update-ca-trust` (FIXME: do we need `extract` or not?) at start.
+They must also run `update-ca-trust` periodically thereafter to pull in any [updates from the mounted ConfigMap][mounted-configmap-updates].
+
+Components that need a single trust bundle injected can bypass `update-ca-trust` and mount their `ca-bundle.crt` directly into `/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem` (FIXME: RHEL/UBI docs confirming this as a supported workflow).
```

Comment:
Could we automate the createion of environment variables containing the cert bundles, or create a volume using [the same mechanism as proxy](https://github.com/openshift/enhancements/blob/19d5bc00db44cf2614634e7e09e28743b6d87a62/enhancements/proxy/global-cluster-egress-proxy.md#implementation-detailsnotesconstraints-optional). It would streamline how we use the injection process.

Reply:
> Could we automate the creation of environment variables containing the cert bundles...

It's not clear to me how this would work.  Can you give example Deployment YAML and talk us through this?

Reply:
Sure, I was just extrapolating from this excerpt from the [global cluster egress policy](https://github.com/openshift/enhancements/blob/19d5bc00db44cf2614634e7e09e28743b6d87a62/enhancements/proxy/global-cluster-egress-proxy.md#implementation-detailsnotesconstraints-optional) enhancement. 

> deployments with the config.openshift.io/inject-proxy: <container-name> will get the current proxy environment variables injected (HTTP_PROXY, HTTPS_PROXY, NO_PROXY) by the cluster version operator.

So in our case we could add `config.openshift.io/add-trusted-ca-bundle: <container name>` to our deployments to automatically mount the trust-bundle config map to a volume, then add an inline command to that container to add it to its CA trusts.

Reply:
@iamemilio I'm happy that I came across your comment. I'm working on adding proxy support to the ingress operator  and was considering a similar approach. I feel like we need a better solution to mounting the trust bundle than what currently exists (see [image registry operator](https://github.com/openshift/cluster-image-registry-operator/commit/8b59dceb5a690d93f592cb83a3d0fc6359437103)).

@wking let's use ingress operator as an example. The [deployment](https://github.com/openshift/cluster-ingress-operator/blob/master/manifests/02-deployment.yaml) would be updated with:
```
  annotations:
    config.openshift.io/add-trusted-ca-bundle: "ingress-operator" # container name to inject bundle
```

CVO watches deployments with the above annotation and creates a volume for the operator's trust bundle configmap and mounts the volume in the container(s) specified by the annotation. The ingress operator's deployment would look like the following after CVO finishes reconciliation:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ingress-operator
  namespace: openshift-ingress-operator
  annotations:
    config.openshift.io/add-trusted-ca-bundle: ingress-operator
spec:
<SNIP>
      containers:
        - name: ingress-operator
<SNIP>
          volumeMounts:
            - name: trusted-ca
              mountPath: /var/run/configmaps/trusted-ca/
<SNIP>
        configMap:
          name: trusted-ca
          optional: true
          items:
            - key: ca-bundle.crt
              path: tls-ca-bundle.pem
```

Reply:
i think you will have a hard time making the case that the CVO should do this.

the only reason the CVO took ownership of setting env vars on the operator deployments is because it is impossible for the operator to set env vars on its own deployment(the cvo would stomp away the values).  The same restriction doesn't apply to the configmap volume mounting, since the operator can define its deployment with a CM volume mount, and define the CM to be mounted.

Reply:
@bparees thanks for the feedback. I'll look into having the ingress operator programmatically manage it's own trust bundle volume/volumeMount. However, having a single controller that manages this would provide consistency across operators.

Reply:
@bparees what would you suggest as a solution then? Should an operator be given the responsibility for injecting this into container deployments, or do you think we should drop it an just add it manually to all the services.

Reply:
@iamemilio today operators that need proxy CAs in their operator deployment, configure themselves accordingly:

1) they define a configmap that will have the CAs injected into it:
https://github.com/openshift/cluster-image-registry-operator/blob/a0efd48f8bc2c8d18d53fc5a9e5b8515b94bc2a4/manifests/04-ca-trusted.yaml#L8

2) they mount that configmap in their deployment:
https://github.com/openshift/cluster-image-registry-operator/blob/a0efd48f8bc2c8d18d53fc5a9e5b8515b94bc2a4/manifests/07-operator.yaml#L93

3) they watch the mounted content for changes and restart their process as needed:
https://github.com/openshift/cluster-image-registry-operator/blob/a0efd48f8bc2c8d18d53fc5a9e5b8515b94bc2a4/manifests/07-operator.yaml#L62-L73

That is what i would expect here